plugins/modules/ec2_metadata_facts.py

```diff
@@ -484,11 +487,47 @@ def __init__(
         self._token = None
         self._prefix = "ansible_ec2_%s"
 
+    def decode_user_data(self, data):
+        is_compressed = False
+
+        # Get the current locale's encoding
+        current_encoding = locale.getpreferredencoding()
+
+        # Check if data is compressed using zlib header
+        if data.startswith(b'\x78\x9c') or data.startswith(b'\x1f\x8b'):
+            is_compressed = True
+
+        if is_compressed:
+            # Data is compressed, decompress and decode
+            try:
+                # Decompress using zlib
+                decompressed_data = zlib.decompress(data, zlib.MAX_WBITS | 32).decode(current_encoding)
+            except zlib.error:
+                # Unable to decompress, return original data
+                return data.decode(current_encoding)
+
+            try:
+                # Attempt to decode using base64
+                decoded_data = base64.b64decode(decompressed_data)
+                return decoded_data.decode(current_encoding)
+            except:
+                # Unable to decode using base64, return as bytes
+                return decompressed_data
+        else:
+            # Data is not compressed, decode data using base64
+            try:
+                decoded_data = base64.b64decode(data)
+                return decoded_data.decode(current_encoding)
```

Comment:
The documentation states that the user data is already base64 decoded when you retrieve it from instance metadata, so this shouldn't be necessary.